File: webRobot22/src/main/java/connQak/ConnQakBase.java
package connQak;

import unibo.actor22comm.ProtocolType;
import unibo.actor22comm.interfaces.Interaction2021;
import unibo.actor22comm.utils.ColorsOut;

public abstract class ConnQakBase  {

    private static ConnQakBase currQakConn;
    public static final String ctxqakdest      = "ctxbasicrobot";
    public static final String qakdestination 	= "basicrobot";

    public static ConnQakBase create(ProtocolType protocol) {
        ColorsOut.outappl("ConnQakBase | create protocol="+protocol, ColorsOut.MAGENTA);

        if( protocol == ProtocolType.tcp ){
            currQakConn = new ConnQakTcp( );
            return currQakConn;
        } else if( protocol == ProtocolType.coap ){
            currQakConn = new ConnQakCoap( );
            return currQakConn;
        }
        else return null;
    }

    public abstract Interaction2021 createConnection(String host, int port  );
    public abstract void forward(String msg);
    public abstract void request(String msg);
    public abstract void emit(String msg);



}

File: it.unibo.qak21.robots/resources/connQak/connQakBase.kt
package connQak 
import it.unibo.`is`.interfaces.IObserver
import java.util.Observable
import it.unibo.kactor.ApplMessage
 
enum class ConnectionType {
    TCP, MQTT, COAP, HTTP
}

abstract class connQakBase() {
lateinit var currQakConn  : connQakBase
	
	companion object{
	fun create(connType: ConnectionType) : connQakBase{
		  showSystemInfo()
		  when( connType ){
				 ConnectionType.MQTT ->  {return connQakMqtt( )}  
				 ConnectionType.TCP ->   {return connQakTcp( ) }  
				 ConnectionType.COAP ->  {return connQakCoap( )}  
  				 ConnectionType.HTTP ->  {return connQakHttp( )} 
//   				 else -> //println("WARNING: protocol unknown")
 		  }		
	}
	fun showSystemInfo(){
		println(
			"connQakBase  | COMPUTER memory="+ Runtime.getRuntime().totalMemory() +
					" num of processors=" +  Runtime.getRuntime().availableProcessors());
		println(
			"connQakBase  | NUM of threads="+ Thread.activeCount() +
					" currentThread=" + Thread.currentThread() );
	}
	}//object

	
	  abstract fun createConnection( )     
      abstract fun forward( msg : ApplMessage )
      abstract fun request( msg : ApplMessage )
      abstract fun emit( msg : ApplMessage )
	
}

 
 
 